### PR TITLE
Update the placement section of the policy CR

### PR DESCRIPTION
**Pull request by clyang82**
update the placement section of the policy status CR. it can be calculated from the placementbinding in HoH cluster.

Signed-off-by: clyang82 <chuyang@redhat.com>

### CODE REVIEW

File: pkg/dbsyncers/policy_syncer.go [hunk 1]
```diff
@@ -149,10 +151,87 @@ func updateComplianceStatus(ctx context.Context, k8sClient client.StatusClient,
...
+
+	}
+	return placements, nil
+}
+
+func getPlacementBindingList(k8sClient client.Client, instance *policiesv1.Policy) (*policiesv1.PlacementBindingList, error) {
+	// Get the placement binding in order to later get the placements
+	pbList := &policiesv1.PlacementBindingList{}
+
+	err := k8sClient.List(context.TODO(), pbList, &client.ListOptions{Namespace: instance.GetNamespace()})
```
Comment (nirrozenbaum) [inline pkg/dbsyncers/policy_syncer.go]: context should be passed as argument from upper level function.
if the controller is stopped from some reason (e.g. rolling upgrade as part of version change), the context from the upper level is cancelled and then we have graceful shutdown.  

we documented this best practice here - https://github.com/stolostron/hub-of-hubs/blob/main/development.md#context-usage

File: pkg/dbsyncers/policy_syncer.go [hunk 2]
```diff
@@ -149,10 +151,87 @@ func updateComplianceStatus(ctx context.Context, k8sClient client.StatusClient,
...
+	}
+	return pbList, nil
+}
+
+// getPlacementsPerBinding returns the placements for the policy per placementbinding
+func getPlacementsPerBinding(k8sClient client.Client, pb policiesv1.PlacementBinding,
+	instance *policiesv1.Policy) ([]*policiesv1.Placement, error) {
+	plr := &placementrulesv1.PlacementRule{}
+
+	err := k8sClient.Get(context.TODO(), types.NamespacedName{
```
Comment (nirrozenbaum) [inline pkg/dbsyncers/policy_syncer.go]: same comment about context as previous one

File: pkg/dbsyncers/policy_syncer.go [hunk 3]
```diff
@@ -149,10 +151,87 @@ func updateComplianceStatus(ctx context.Context, k8sClient client.StatusClient,
...
+		Namespace: instance.GetNamespace(),
+		Name:      pb.PlacementRef.Name,
+	}, plr)
+	// no error when not found
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get placementrule: %w", err)
+	}
+
```
Comment (nirrozenbaum) [inline pkg/dbsyncers/policy_syncer.go:222]: add return for `errros.IsNotFound(err)=true` case

File: pkg/dbsyncers/policy_syncer.go [hunk 4]
```diff
@@ -149,10 +151,87 @@ func updateComplianceStatus(ctx context.Context, k8sClient client.StatusClient,
...
+	for _, subject := range pb.Subjects {
+		if subject.Kind == policiesv1.Kind && subject.Name == instance.GetName() && !plcPlacementAdded {
+			placement := &policiesv1.Placement{
+				PlacementBinding: pb.GetName(),
+				PlacementRule:    plr.GetName(),
+			}
+			placements = append(placements, placement)
+			// should only add policy placement once in case placement binding subjects contains duplicated policies
+			plcPlacementAdded = true
```
Comment (nirrozenbaum) [inline pkg/dbsyncers/policy_syncer.go]: why not using `break` instead of this boolean?
e.g. something like:
```
if subject.Kind != policiesv1.Kind || subject.Name != instance.GetName() {
    continue
}
// kind and name match the policy
placement := &policiesv1.Placement{
	PlacementBinding: pb.GetName(),
	PlacementRule:    plr.GetName(),
}
placements = append(placements, placement)
break // should only add policy placement once in case placement binding subjects contains duplicated policies
```

File: pkg/dbsyncers/policy_syncer.go [hunk 5]
```diff
@@ -24,12 +25,13 @@ const (
 	complianceStatusTableName = "compliance"
 )
 
+var log = ctrl.Log.WithName("policies-db-syncer")
```
Comment (nirrozenbaum) [inline pkg/dbsyncers/policy_syncer.go]: linters don't like global variables. global variable violates GoLang best practices (unless it's err definions)
this usually gives linter errors and reduces the code quality when doing static analysis in code check tools.
e.g. - https://goreportcard.com/report/github.com/stolostron/hub-of-hubs-status-sync

in addition, this log variable is now available in all dbsyncers.
this may cause human errors when other developers work on this code and they might use the global log and we may get wrong logging, which I guess is not what you meant.

File: pkg/dbsyncers/policy_syncer.go [hunk 6]
```diff
@@ -149,10 +151,87 @@ func updateComplianceStatus(ctx context.Context, k8sClient client.StatusClient,
...
 	if err != nil && !errors.IsNotFound(err) {
 		return fmt.Errorf("failed to update policy CR: %w", err)
 	}
 
 	return nil
 }
+
+func getPlacements(k8sClient client.Client, instance *policiesv1.Policy) ([]*policiesv1.Placement, error) {
+	log.Info("get the placements for the policy", "policy", instance.GetName())
+	pbList, err := getPlacementBindingList(k8sClient, instance)
```
Comment (nirrozenbaum) [inline pkg/dbsyncers/policy_syncer.go]: this implementation is very wasteful.
think about the following example - 
let's assume we have 100 policies in the same namespace. this code will get all the bindings in the namespace 100 times.
that means that if we have 100 bindings, we get each 100 times - the code will get bindings 10000 times with no good reason (!).

I suggest to change this. we may use a cache in memory for bindings per namespace (map from ns -> list of bindings).
when we handle a policy we can check in the cache if we already read the bindings in the ns and then we don't need to get it again. in every sync cycle we can build the cache from scratch for simplicity.  

this will reduce the number of bindings we get in the above example to 100.

File: pkg/dbsyncers/policy_syncer.go [hunk 7]
```diff
@@ -149,10 +151,87 @@ func updateComplianceStatus(ctx context.Context, k8sClient client.StatusClient,
...
+	var placements []*policiesv1.Placement
+	for _, pb := range pbList.Items {
+		subjects := pb.Subjects
+		for _, subject := range subjects {
+			if !(subject.APIGroup == policiesv1.SchemeGroupVersion.Group &&
+				subject.Kind == policiesv1.Kind &&
+				subject.Name == instance.GetName()) {
+				continue
+			}
+			p, err := getPlacementsPerBinding(k8sClient, pb, instance)
```
Comment (nirrozenbaum) [inline pkg/dbsyncers/policy_syncer.go]: can you use longer and more descriptive names? we have now variables `p`, `pb`, `plr`, `pbList`...
this code gets very confusing with these one/two char names and gets hard to follow.

can we call the variables with a name that describes them?
e.g. `placementRule`, `placementBinding`, `placementBindingList`, etc.

**Review by nirrozenbaum — COMMENTED** [review]
@clyang82 overall looks good.
I wrote some comments.

File: pkg/dbsyncers/policy_syncer.go (hunk 1 above)
Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go]: yes. I forgot to pass it.

File: pkg/dbsyncers/policy_syncer.go (hunk 3 above)
Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go:222]: yes. no need to handle if the error is not found or found. just return error if the err is not nil.

File: pkg/dbsyncers/policy_syncer.go (hunk 4 above)
Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go]: Yes. that makes sense. the `PlacementBinding` can refer `placementrule` to multiple policies. but here we only handle one policy. so `break` is good. Thanks.

File: pkg/dbsyncers/policy_syncer.go (hunk 5 above)
Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go]: Yes. that is not my intention. Thanks. I do not like to pass the `log` as parameter one by one. we may consider to put `log` into `genericDBSyncer` struct. WDYT?

File: pkg/dbsyncers/policy_syncer.go (hunk 6 above)
Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go]: I agree with your point that this implementation is very wasteful. That is existing logic in `governance-policy-propagator`. I am thinking it carefully. I still think if the policy is a global policy, then it should be handled by HoH controller. for the long term, the hoh controller can watch the placementbinding to update the policy.  we need to introduce the mechanism to update the cache if we cache the placementbindings.

for the short term, just use this solution to update the placements section. WDYT?

File: pkg/dbsyncers/policy_syncer.go (hunk 3 above)
Comment (nirrozenbaum) [reply pkg/dbsyncers/policy_syncer.go:222]: what do you mean no need to handle this case?
if `errors.IsNotFound(err) = true`, that means the code wasn't able to read the required placementrule and then the  variable from line 212 stays empty.
if we continue the function without return we will update all subjects with empty PlacementRule (placementrule.GetName() will return an empty string in line 229)

File: pkg/dbsyncers/policy_syncer.go (hunk 5 above)
Comment (nirrozenbaum) [reply pkg/dbsyncers/policy_syncer.go]: I agree. putting log (and also k8sClient I think) makes perfect sense to go into the generic db syncer.
at the end, every db syncer will use log + client for sure. 

minor changes are required for this. let's finish this PR while keeping things this way and then I'll change code to work with log + k8sClient inside the generic db syncer.

File: pkg/dbsyncers/policy_syncer.go (hunk 6 above)
Comment (nirrozenbaum) [reply pkg/dbsyncers/policy_syncer.go]: I'm not sure this will work good.
we want status sync to be responsive and therefore we configured it to have a sync interval of 5 seconds (5 seconds between consecutive syncs).

if we're talking about few bindings, it will probably work fine.
if we take the above example - there is no chance that status sync will manage to read 10K resources in 5 seconds cycle and we might get to a non responsive status sync (if sync didn't finish in the interval context might get cancelled and nothing is updated. 
see here - https://github.com/stolostron/hub-of-hubs-status-sync/blob/main/pkg/dbsyncers/generic_db_syncer.go#L52-L54)

**Review by nirrozenbaum — COMMENTED** [review]
@clyang82 wrote few comments

File: pkg/dbsyncers/policy_syncer.go (hunk 3 above)
Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go:222]: I mean that if there is error, then return. otherwise, the placementrule is not empty.

File: pkg/dbsyncers/policy_syncer.go (hunk 6 above)
Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go]: I noticed this problem existed in the current implementation. any concerns to add the logic to update placement status in policy CR in spec-sync? in spec-sync, we watch the placementbinding, so I think that is a proper place to do it. And we can keep the policy spec to have placement status only in CR and DB. I mean do not clean up the `placement` in status. we can keep it to sync-up to the leaf hub. It should not any impacts.

Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go]: I do not want to put it in governance-policy-propagator due to:
1. if the policy is global policy, it should be handled by HoH controller.
2. It can be easy to adopt by policy squad via filtering out the global policy in mapper. e.g.: https://github.com/clyang82/governance-policy-propagator/blob/5844672aff73fe4915ae54020a3ca5d32600296a/controllers/propagator/policyMapper.go#L58
Does it make sense? @nirrozenbaum

File: pkg/dbsyncers/policy_syncer.go (hunk 3 above)
Comment (nirrozenbaum) [reply pkg/dbsyncers/policy_syncer.go:222]: got it. sorry I missed the removal of the `IsNotFound` check.

File: pkg/dbsyncers/policy_syncer.go (hunk 6 above)
Comment (nirrozenbaum) [reply pkg/dbsyncers/policy_syncer.go]: yes. it makes sense.
let's take the option to put it in policy propagator off the table.
so we're left only with the option to handle global policies in one of the HoH controllers.

regarding the above suggestion - this violates the separation of concerns we kept in the design and is mixing spec/status updates. I don't think this is a good idea.
the clean separation allows us flexibility and customization that this change may break.
let me think about alternative approach.

Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go]: @nirrozenbaum please take a look at this approach - https://github.com/stolostron/hub-of-hubs-manager/pull/15

Comment (nirrozenbaum) [reply pkg/dbsyncers/policy_syncer.go]: @clyang82 this approach makes a lot of sense.
+1 from me on this.

wrote two minor comments in hub-of-hubs-manager PR.

Comment (clyang82) [reply pkg/dbsyncers/policy_syncer.go]: Thanks @nirrozenbaum let me close this PR. trace it in https://github.com/stolostron/hub-of-hubs-manager/pull/15
